eth/executionclient/execution_client.go

```diff
@@ -31,76 +32,164 @@ var (
 
 const elResponseErrMsg = "Execution client returned an error"
 
-// ExecutionClient represents a client for interacting with Ethereum execution client.
+// ExecutionClient represents a client for interacting with Ethereum execution clients.
 type ExecutionClient struct {
 	// mandatory
-	nodeAddr        string
+	nodeAddrs       []string
 	contractAddress ethcommon.Address
 
 	// optional
 	logger *zap.Logger
 	// followDistance defines an offset into the past from the head block such that the block
 	// at this offset will be considered as very likely finalized.
-	followDistance              uint64 // TODO: consider reading the finalized checkpoint from consensus layer
+	followDistance              uint64
 	connectionTimeout           time.Duration
 	reconnectionInitialInterval time.Duration
 	reconnectionMaxInterval     time.Duration
 	logBatchSize                uint64
 
 	syncDistanceTolerance uint64
-	syncProgressFn        func(context.Context) (*ethereum.SyncProgress, error)
+	syncProgressFn        func(context.Context, *ManagedClient) (*ethereum.SyncProgress, error)
 
 	// variables
-	client *ethclient.Client
-	closed chan struct{}
+	clientsMu   sync.RWMutex
+	clients     []*ManagedClient
+	closed      chan struct{}
+	closingOnce sync.Once
 }
 
 // New creates a new instance of ExecutionClient.
 func New(ctx context.Context, nodeAddr string, contractAddr ethcommon.Address, opts ...Option) (*ExecutionClient, error) {
+	addrList := strings.Split(nodeAddr, ";")
+	if len(addrList) == 0 {
+		return nil, fmt.Errorf("no node address provided")
+	}
```

Comment:
as I said above, I think it makes more sense to pass a slice. and to the parsing on a higher level.

Reply:
done